packages/compiler/src/render3/partial/component.ts

```diff
@@ -0,0 +1,109 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import * as core from '../../core';
+import {DEFAULT_INTERPOLATION_CONFIG} from '../../ml_parser/interpolation_config';
+import * as o from '../../output/output_ast';
+import {Identifiers as R3} from '../r3_identifiers';
+import {R3ComponentMetadata, R3DirectiveDef} from '../view/api';
+import {createComponentType} from '../view/compiler';
+import {ParsedTemplate} from '../view/template';
+import {DefinitionMap} from '../view/util';
+
+import {createDirectiveDefinitionMap} from './directive';
+import {toOptionalLiteralArray} from './util';
+
+
+/**
+ * Compile a directive declaration defined by the `R3DirectiveMetadata`.
+ */
+export function compileDeclareComponentFromMetadata(
+    meta: R3ComponentMetadata, template: ParsedTemplate): R3DirectiveDef {
+  const definitionMap = createComponentDefinitionMap(meta, template);
+
+  const expression = o.importExpr(R3.declareComponent).callFn([definitionMap.toLiteralMap()]);
+  const type = createComponentType(meta);
+
+  return {expression, type};
+}
+
+/**
+ * Gathers the declaration fields for a component into a `DefinitionMap`.
+ */
+export function createComponentDefinitionMap(
+    meta: R3ComponentMetadata, template: ParsedTemplate): DefinitionMap {
+  const definitionMap = createDirectiveDefinitionMap(meta);
+
+  const templateMap = new DefinitionMap();
+  templateMap.set(
+      'source',
+      typeof template.template === 'string' ? o.literal(template.template) : template.template);
+  templateMap.set('isInline', o.literal(template.isInline));
+
+  definitionMap.set('template', templateMap.toLiteralMap());
```

Comment:
NIT: I would move this into its own helper function - to keep this function nice and simple:

```suggestion
  definitionMap.set('template', getTemplateDefinition(template));
```